pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,308 @@
+	c.Lock()
+	defer c.Unlock()
+
+	scData, ok := c.statusCollectorNameToData[statusCollectorName]
+	if !ok {
+		return false // statusCollector is irrelevant to this combinedstatus resolution
+	}
+
+	if scData.equals(clauses) {
+		return false // statusCollector data is already up-to-date
+	}
+
+	// status collector clauses need to be updated, therefore update fields
+	// and invalidate all cached workstatus evaluations by resetting the map
+	c.statusCollectorNameToData[statusCollectorName] = &statusCollectorData{
+		statusCollectorClauses: clauses,
+		workStatusRefToData:    make(map[util.ObjectIdentifier]workStatusData),
+	}
+
+	return true
+}
+
+// evaluateWorkStatus evaluates the workstatus per relevant statuscollector.
+// The function returns true if any statuscollector data is updated.
+func (c *combinedStatusResolution) evaluateWorkStatus(ctx context.Context, celEvaluator *celEvaluator,
+	workStatusIdentifier util.ObjectIdentifier, workStatusContent *runtime.RawExtension) (bool, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	logger := klog.FromContext(ctx)
+
+	updated := false
+	for statusCollectorName := range c.statusCollectorNameToData {
+		scData, ok := c.statusCollectorNameToData[statusCollectorName]
+		if !ok {
+			continue
+		}
+
+		changed, err := c.evaluateWorkStatusAgainstStatusCollector(celEvaluator, workStatusIdentifier,
+			workStatusContent, scData)
+		if err != nil {
+			logger.Error(err, "failed to evaluate workstatus against statuscollector",
+				"statusCollectorName", statusCollectorName)
+			return false, nil
+		}
+
+		updated = updated || changed
+	}
+
+	return updated, nil
+}
+
+// evaluateWorkStatusAgainstStatusCollector evaluates the workstatus against
+// the statuscollector clauses and caches the evaluations. The function returns
+// true if an evaluation is updated.
+// If any evaluation fails, the function returns an error.
+func (c *combinedStatusResolution) evaluateWorkStatusAgainstStatusCollector(celEvaluator *celEvaluator,
+	workStatusIdentifier util.ObjectIdentifier, workStatusContent *runtime.RawExtension,
+	scData *statusCollectorData) (bool, error) {
+	wsData, ok := scData.workStatusRefToData[workStatusIdentifier]
+	if !ok {
+		wsData = workStatusData{}
+		scData.workStatusRefToData[workStatusIdentifier] = wsData
+	}
+
+	updated := false
+	// evaluate all clauses
+	if scData.filter != nil {
+		eval, err := celEvaluator.Evaluate(string(*scData.filter), workStatusContent)
+		if err != nil {
+			return false, fmt.Errorf("failed to evaluate filter expression: %w", err)
+		}
+
+		if eval.Type().TypeName() != "bool" {
+			return false, fmt.Errorf("filter expression must evaluate to a boolean")
+		}
+
+		fiterEval := eval.Value().(bool)
+		updated = fiterEval != wsData.filterEval
+		wsData.filterEval = fiterEval
+	}
+
+	// evaluate select
+	selectEvals := make(map[string]any)
+	for _, selectNamedExp := range scData.selectClause {
+		eval, err := celEvaluator.Evaluate(string(selectNamedExp.Def), workStatusContent)
```

Comment:
Maybe redefine `celEvaluator.Evaluate` so that every call site does not need a cast?